File: benchmark/autolifter/single-pass/3rd-min.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if > a b then a else b;

trdmin = \xs: List. (fix (
  \f: List -> {Int, Int, Int}. \xs: List.
  match xs with
    nil _ -> {inf, inf, inf}
  | cons {h, t} ->
    let res = f t in
      {min res.1 (max res.2 h),
       min res.2 (max res.3 h),
       min res.3 h}
  end
) xs).1;

main = single_pass trdmin;

File: ml-benchmark/autolifter/single-pass/3rd-min.ml
type tlist = Cons of { h: int; t: tlist } | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons {h; t} -> Cons {h; t = run t}
  in
  fun xs -> v (run xs)

let inf = 100

let min a b = if a < b then a else b
let max a b = if a > b then a else b

let trdmin xs =
  let rec f xs =
    match xs with
    | Nil _ -> (inf, inf, inf)
    | Cons {h; t} ->
        let res = f t in
        match res with
        | (r1, r2, r3) ->
            (min r1 (max r2 h), min r2 (max r3 h), min r3 h)
  in
  match f xs with
  | (a, _, _) -> a

let program = single_pass trdmin